plenum/server/node.py

```diff
@@ -2662,20 +2674,28 @@ def sendNodeRequestSpike(self):
 
     def primary_selected(self, instance_id):
         # If the node has primary replica of master instance
-        if instance_id == 0:
+        if instance_id == self.master_replica.instId:
             # TODO: 0 should be replaced with configurable constant
             self.monitor.hasMasterPrimary = self.has_master_primary
-        if not self.lost_primary_at:
-            return
-        if self.nodestack.isConnectedTo(self.master_primary_name) or \
-                self.master_primary_name == self.name:
-            self.lost_primary_at = None
+            if not self.primaries_disconnection_times[self.master_replica.instId]:
+                return
+            if self.nodestack.isConnectedTo(self.master_primary_name) \
+                    or self.master_primary_name == self.name:
+                self.primaries_disconnection_times[self.master_replica.instId] = None
+        else:
+            primary_node_name = self.replicas[instance_id].primaryName.split(':')[0]
+            if self.nodestack.isConnectedTo(primary_node_name) \
+                    or primary_node_name == self.name:
+                self.primaries_disconnection_times[instance_id] = None
+            else:
```

Comment:
View change completion is the moment when replicas get to know their primaries. At this moment the node should check the current statuses of connections to the primaries (since the node may already be disconnected from another node which becomes the primary in some instance in the new view). For the master replica this check is redundant because if its primary were disconnected then the node would not receive `ViewChangeDone` message and thus would not complete the view change. But for backup replicas this check on view change completion is required. It is also used when the node starts (completes primaries selection for the view 0).